deepspeed/runtime/zero/stage2.py

```diff
@@ -1376,22 +1383,25 @@ def buffered_reduce_fallback(self,
 
     # views the tensor as multiple partitions and returns
     # those partitions
-    def get_data_parallel_partitions(self, tensor, group_id):
+    def get_data_parallel_partitions(self, tensor, group_id, alignment_factor):
         partitions = []
 
         dp = dist.get_world_size(group=self.real_dp_process_group[group_id])
         dp_id = dist.get_rank(group=self.real_dp_process_group[group_id])
 
         total_num_elements = tensor.numel()
 
-        base_size = total_num_elements // dp
-        remaining = total_num_elements % dp
+        divisor = dp * alignment_factor
+        # ceil( total_num_elements / divisor ) using integers
+        base_size = ( total_num_elements // divisor ) \
+            + ( total_num_elements % divisor > 0)
+        base_size *= alignment_factor
 
         start = 0
         for id in range(dp):
             partition_size = base_size
-            if id < remaining:
-                partition_size = partition_size + 1
+            if id == (dp - 1):
```

Comment:
It seems to me that the new code is not functionality equivalent even if alignment_factor = 1 
Consider tne=6, dp=4, where tne is total_num_elements
In the old code, we have 
tne/dp = 1
tne%dp = 2
partition_size = [2, 2, 1, 1]

In the new code, we have partition_size = [2,2,2,0]

Am I correct?

Reply:
Yes. That would be correct. 

The tne is a big number usually. 
The particular test we were looking for in MSRCHA-137 (run_glue) had tne=1,566,913,544. Original base_size was 195864193, across 8-gpus. This particular change would make base_size = 195864194. 
The partitions then are [195864194, 195864194, ... , (195864194-8)]. The last partition can go up to (base_size-15), where 15 = (alignment_factor*num_gpus-1).

Reply:
Thanks for the response and context. We need this to work correctly even for small sizes of tne, and I suspect you will run into such cases in our unit tests. I think it is fine to pad total_num_elements for cases where total_num_elements < alignment_factor * dp, and your change [here](https://github.com/amathews-amd/DeepSpeed/blob/a0d99dbdcefc7fd8d41435a21f7d3ac084b038ad/deepspeed/runtime/zero/stage2.py#L291-L292) seems to do that already.  However, rather than align to (alignment_factor * dp_size) which could lead to excessive padding, why not align to LCM(alignment_factor, dp_size)?

Reply:
LCM(alignment_factor, dp_size) would not work. We split the partitions based on dp_size already in both the original code and new case. After this division, we need the first element of each partition to have a 4-byte position. For FP16, this means the address of starting element of each partition has to be multiple of 2.

Reply:
Yes, you are correct. I was wrong. Thanks for the explanation.